From a0056ea1d994f1ec4da8ccae45abab2d2461f0a2 Mon Sep 17 00:00:00 2001
From: Gyorgy Szing <Gyorgy.Szing@arm.com>
Date: Thu, 16 Nov 2023 18:14:46 +0000
Subject: [PATCH 1/1] smm_gateway: GetNextVariableName Fix

GetNextVariableName() should return EFI_BUFFER_TOO_SMALL
when requested NameSize is smaller than the actual. It
currently returns EFI_BUFFER_OUT_OF_RESOURCES due to setting
max_name_len incorrectly. This change fixes the error by
using clamping the maximum size to the NameSize requested by
the client.

Upstream-Status: Pending
Signed-off-by: Emekcan Aras <emekcan.aras@arm.com>
Signed-off-by: Gyorgy Szing <Gyorgy.Szing@arm.com>
---
 .../service/smm_variable/provider/smm_variable_provider.c     | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/components/service/smm_variable/provider/smm_variable_provider.c b/components/service/smm_variable/provider/smm_variable_provider.c
index f1c3c712..7ec49af5 100644
--- a/components/service/smm_variable/provider/smm_variable_provider.c
+++ b/components/service/smm_variable/provider/smm_variable_provider.c
@@ -190,15 +190,13 @@ static rpc_status_t get_next_variable_name_handler(void *context, struct rpc_req
 		if (resp_buf->size >= param_len) {
 
 			struct rpc_buffer *req_buf = &req->request;
-			size_t max_name_len = resp_buf->size -
-				SMM_VARIABLE_COMMUNICATE_GET_NEXT_VARIABLE_NAME_NAME_OFFSET;
 
 			memmove(resp_buf->data, req_buf->data, param_len);
 
 			efi_status = uefi_variable_store_get_next_variable_name(
 				&this_instance->variable_store,
 				(SMM_VARIABLE_COMMUNICATE_GET_NEXT_VARIABLE_NAME*)resp_buf->data,
-				max_name_len,
+				((SMM_VARIABLE_COMMUNICATE_GET_NEXT_VARIABLE_NAME*)resp_buf->data)->NameSize,
 				&resp_buf->data_length);
 		}
 		else {
-- 
2.34.1

